src/db-adapters/Mongoose/MongooseAdapter.js

```diff
@@ -603,27 +604,41 @@ export default class MongooseAdapter {
     return words.join(" ");
   }
 
-  static getIdQueryType(idOrIds) {
-    const mode = typeof idOrIds === "string" ? "findOne" : "find";
-    let idQuery;
+  static getIdQueryType(idOrIds, model) {
+    return Q.Promise((resolve, reject) => {
+      const mode = typeof idOrIds === "string" ? "findOne" : "find";
+      let ids = Array.isArray(idOrIds) ? idOrIds : [ idOrIds ];
+      let idQuery;
+      let tests = [];
 
-    if (typeof idOrIds === "string") {
-      if (!this.idIsValid(idOrIds)) {
-        throw new APIError(404, undefined, "No matching resource found.", "Invalid ID.");
-      }
+      ids = ids.filter(id => Boolean(id));
```

Comment:
Good call. But I think that case only applies if a single id is passed in. (I.e., if `idOrIds` is `[null]`, that should be an error, right?) So I think I'd replace the filter with a conditional that only checks when a single ids is passed in. (I'm not sure what the resolution there should be though; see comment.) 

Also, it'd be nice to clarify that `mode` and `ids` are based on the same isArray check. Finally, I think `idIsValid` should resolve to true, not to the id, when the id is valid (since it has "is" in the name). 

Putting those together:

```
 static getIdQueryType(idOrIds) {
    const [mode, ids] =  Array.isArray(idOrIds) ? ["find", idOrIds] : ["findOne", [idOrIds]];
    
    if(mode === "findOne" && idOrIds == null) {
      // what should this resolution be? 
      // ["findOne", {_id: undefined}] or ["findOne", undefined]
      // I think it was the form before this pr, but is now the latter
    }
    
    const idValidationPromises = ids.map(id => this.idIsValid(id, model));

   return Q.all(idValidationPromises).then(() => {
     const idQuery = { _id: mode === "find" ? { $in: ids } : ids[0] };
     return [ mode, idQuery];
   }, (err) => {
     // ....
   });
});